dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/VisitorInlineAsm.java

```diff
@@ -0,0 +1,619 @@
+        Label label = getOrNewLabel(labelID);
+        asmInstructions.add(label);
+        return null;
+    }
+
+    @Override
+    public Object visitExpr(InlineAsmParser.ExprContext ctx) {
+        return ctx.getChild(0).accept(this);
+    }
+
+    // If the register with that ID was already defined, we simply return it
+    // otherwise, we create and return the new register.
+    // each time we call this visitor, it picks up the ID of the register -- e.g. $3 -> ID = 3.
+    // if we created a register which refers to the return Register, we have to add to "pendingRegisters", 
+    // as we are going to need it during visitMetadata part
+    @Override
+    public Object visitRegister(InlineAsmParser.RegisterContext ctx) {
+        String registerName = ctx.NumbersInline().getText();
+        int registerID = Integer.parseInt(registerName);
+        // if w.h. the register, return it
+        if (asmRegisters.containsKey(registerID)) {
+            return asmRegisters.get(registerID);
+        } else {
+            // otherwise, pick up the correct type and create the new Register
+            Type registerType = getLlvmRegisterTypeGivenAsmRegisterID(registerID);
+            Register newRegister = this.llvmFunction.getOrNewRegister(makeRegisterName(registerID), registerType);
+            // if the register is part of the return register, we have to register it for the final assignment for the output
+            // as we are going to do it later when we have enough information in visitAsmMetadata when visiting the constraints.
+            if (isPartOfReturnRegister(registerID) && isReturnRegisterAggregate()) {
+                this.pendingRegisters.add(newRegister);
+            }
+            asmRegisters.put(registerID, newRegister);
+            return newRegister;
+        }
+    }
+
+    // This visitor generates the last two kind of events: 
+    // 1) inputAssignments -> how to map llvm registers (args) to asm ones used in the instructions
+    // 2) outputAssignments -> how to map asm registers to the return register
+    // when we reach this point we have already defined all of the asmInstructions events
+    // which means that we already have created all of the asm registers
+    // we just have to read the constraints, and based on their type, understand if they are going to map
+    // to the (args) llvm registers or to the llvm return Register.
+    @Override
+    public Object visitAsmMetadataEntries(InlineAsmParser.AsmMetadataEntriesContext ctx) {
+        List<InlineAsmParser.ConstraintContext> constraints = ctx.constraint();
+        // we already have all of the registers from LHS of the context call (asmInstrEntries)
+        // now it is time to generate the input and output assignments to let Dat3M understand the llvm to asm registers mapping
+        boolean outputRegistersInitialized = returnRegister == null;
+        for (int i = 0; i < constraints.size(); i++) {
```

Comment:
Add a description along the lines "We iterate until we find the first non-output constraint. Then we immediately initialize the return register (the right-hand side of the assignment will be either a single register or an aggregate type deepening on how many output constraints we processed). We then map args registers to asm registers (we need to shift the register ID to find the corresponding args position of the matching register). Numeric constraint just map the registerID with the corresponding numeric position."

All other comments in the method could be removed.